proposals/scripting-support.md

```diff
@@ -0,0 +1,434 @@
+From an IDE point of view, they are project-context dependent, but may not be part of the project sources. (In the same 
+sense as e.g. gradle build scripts source is not considered as a part of the project sources.)
+
+#### Discovery
+
+The IDE needs to be able to extract scripts environment configurations from the project settings.
+ 
+#### Project-level REPL
+ 
+A REPL that has access to the project's compiled classes. 
+
+## Proposal
+
+### Architecture
+
+The scripting support consists of the following components:
+- **Platform** - interface for compilation and IDE support 
+  - `(scriptSource, scriptDefinition, compilationConfig) -> compiledScript`
+  - `(scriptSource, scriptDefinition, compilationConfig) -> ideDataForRichEditing`
+  - predefined platforms based on the kotlin platforms: /JVM, /JS, /Native
+  - custom/customized implementation possible
+  - compiled scripts cashing should be implemented here
+  - the `compilationConfig` is defined by the platform, and may affect code generation; typically contains:
+    - additional dependencies
+    - *required* scripts
+    - expected parameters
+  - should not keep the state of the script execution
+- **Executor** - the component that receives compiled script instantiates it, and then executes it in a required 
+  environment, supplying any arguments that the script requires:
+  - `(compiledScript, compilationConfig, environment) -> scriptObject`
+  - `(scriptObject, compilationConfig, environment) -> Unit`
+  - uses `compilationConfig` and `environment` for "feeding" the script with the arguments
+  - predefined platform-specific executors available, but usually provided by the scripting host
+  - possible executors
+    - JSR-223
+    - IDEA REPL
+    - Jupyter
+    - Gradle
+    - with specific coroutines context
+    - ...
+- **Preprocessor** - receives the script text before compilation and extracts the `compilationConfig`
+  - `(scriptSource) -> compilationConfig`
+  - *previously called DependenciesResolver*
+  - have access to specific compiler services to:
+    - extract script (file-level) annotations
+    - extract parts of the script (*source sections plugin*)
+    - (potentially) perform script lexing for advanced script analysis
+  - predefined platform-specific preprocessors are available for standard cases, custom one could be provided by the 
+    scripting host
+- **Selector** - receives the script file and/or text before further analysis to determine whether the script belongs
+  to this scripting host and the particular set of components
+  - `(scriptSource) -> Boolean`
+  - required for IDE support
+  - could be combined with the preprocessor, since it may require the same text analysing services 
+  - default and simple implementations are provided
+
+### Script object
+
+Script compiled into the following class
+
+```
+class Script(templateParams..., val implicitReceivers[], additionalArgs: Map<String, Any?>): ScriptDefinition(definitionClassParams...) {
```

Comment:
though - this is an implementation detail, so I'll put it in the same way as `templateParams`